sktime/datatypes/_vectorize.py

```diff
@@ -310,6 +309,118 @@ def _get_item_indexer(self, i: int, X=None):
             col_ind = [col_ind]
         return row_ind, col_ind
 
+    def __iter__(self):
+        """Iterate over all instances.
+
+        Returns
+        -------
+        An iterator over all instances
+        """
+        return (
+            group
+            for _, _, group in self.items(
+                iterate_as=self.iterate_as, iterate_cols=self.iterate_cols
+            )
+        )
+
+    def items(self, iterate_as=None, iterate_cols=None):
+        """Iterate over (group name, column name, instance) tuples.
+
+        Iteration order is "depth first" with columns being the branches and groups the
+        root nodes. Groups are ordered by group keys.
+
+        Row name is null if iterate_as is the same as scitype of data.
+        Col name is null if iterate_cols is False.
+
+        Parameters
+        ----------
+        iterate_as : str ("Series", "Panel"), optional, default=self.iterate_as
+            scitype of the iteration
+            for instance, if X is Panel and iterate_as is "Series"
+                then the class will iterate over individual Series in X
+            or, if X is Hierarchical and iterate_as is "Panel"
+                then the class will iterate over individual Panels in X
+                    (Panel = flat/non-hierarchical collection of Series)
+        iterate_cols : boolean, optional, default=self.iterate_cols
+            whether to iterate over columns, if true, the class will iterate over
+            columns
+
+        Returns
+        -------
+        An iterator over all (row name, column name, instance) tuples.
+        """
+        if iterate_as is None:
+            iterate_as = self.iterate_as
+        self._check_iterate_as(iterate_as)
+
+        if iterate_cols is None:
+            iterate_cols = self.iterate_cols
+        self._check_iterate_cols(iterate_cols)
+
+        def _iter_cols(inst, group_name=None):
+            if iterate_cols:
+                for col in inst.columns:
+                    yield group_name, col, _enforce_index_freq(inst[[col]])
+            else:
+                yield group_name, None, _enforce_index_freq(inst)
+
+        iter_levels = self._iter_levels(iterate_as)
+        is_self_iter = len(iter_levels) == self.X_multiindex.index.nlevels
+
+        if is_self_iter:
+            yield from _iter_cols(self.X_multiindex)
+        else:
+            for name, group in self.X_multiindex.groupby(level=iter_levels):
+                yield from _iter_cols(group.droplevel(iter_levels), group_name=name)
+
+    def _iter_levels(self, iterate_as):
+        """Get the levels to group by for iteration using iterate_as.
+
+        Parameters
+        ----------
+        iterate_as: The scitype of iteration elements
+
+        Returns
+        -------
+        A list of multiindex levels to group by for iteration
+        """
```

Comment:
this could be simplified, no?
```python
iter_levels_dict = {"Panel": 2, "Series": 1}
iter_levels = iter_levels_dict.get(iterate_as, 0)
```
right?

Reply:
(the compatibility between `is_scitype` and `iterate_as` is already checked at construction)

Reply:
I don't think so. For example if `is_scitype == "Series" and iterate_as == "Series"`, `iter_levels` must be 0 not 1.

Reply:
ahhhh, I see. You are right.

How about the following then:

```python
if self.is_scitype != iterate_as:
  iter_levels_dict = {"Panel": 2, "Series": 1}
  iter_levels = iter_levels_dict.get(iterate_as, 0)
else:
  iter_levels = 0
```

Although this seems less explicit and more error prone.

I would say, let's leave it then.